kie-api/src/main/java/org/kie/api/internal/utils/ServiceDiscoveryImpl.java

```diff
@@ -237,4 +259,88 @@ public void put(int priority, K key, V value) {
             return map.entrySet();
         }
     }
-}
+
+    private static Collection<URL> findKieConfUrls(ClassLoader cl) throws IOException {
+        List<URL> kieConfsUrls = new ArrayList<>();
+
+        Enumeration<URL> metaInfs = cl.getResources(CONF_FILE_FOLDER);
+        while (metaInfs.hasMoreElements()) {
+            URL metaInf = metaInfs.nextElement();
+            if (metaInf.getProtocol().startsWith("vfs")) {
+                // the kie.conf discovery mechanism doesn't work under JBoss vfs
+                kieConfsUrls.clear();
+                break;
+            }
+
+            URLConnection con = metaInf.openConnection();
+            if (con instanceof JarURLConnection) {
+                collectKieConfsInJar(kieConfsUrls, metaInf, (JarURLConnection) con);
+            } else {
+                collectKieConfsInFile(kieConfsUrls, new File(metaInf.getFile()));
+            }
+        }
+
+        if (kieConfsUrls.isEmpty()) {
+            // no kie-conf found so fallback to the hardcoded lookup
+            kieConfsUrls = getKieConfsFromKnownModules(cl).collect(Collectors.toList());
+        } else {
+            // check if all discovered kie.conf file are in known modules
+            List<String> notRegisteredModules = kieConfsUrls.stream().map(ServiceDiscoveryImpl::getModuleName)
+                    .filter(module -> Arrays.binarySearch(KIE_MODULES, module) < 0)
+                    .collect(Collectors.toList());
+            if (!notRegisteredModules.isEmpty()) {
+                throw new IllegalStateException("kie.conf file discovered for modules " + notRegisteredModules +
```

Comment:
This will print a list (`notRegisteredModules `). Is `Collectors.toList()` produce a list with pretty print? Otherwise it could make sense to change collect to use a joiner and produce a string